src/zenml/utils/artifact_utils.py

```diff
@@ -393,10 +395,13 @@ def upload_artifact(
                 f"{e}"
             )
 
-    artifact_metadata = {}
+    artifact_metadata: Dict[str, "MetadataType"] = {}
     if extract_metadata:
         try:
             artifact_metadata = materializer.extract_full_metadata(data)
+            step_context = get_step_context()
+            user_provided_metadata = step_context.get_output_metadata(name)
+            artifact_metadata.update(user_provided_metadata)
```

Comment:
Yes correct. I agree there's lot's of room for improvement here.

- API call: this could be optimized for sure. E.g., we could attach all the metadata into the `ArtifactRequestModel` and then let the `create_artifact` call handle the unpacking in the `SQLZenStore`. 
- DB: The main reasons why all values are separate entries are that we wanted to have the step metadata linked to the stack component that created the metadata (even though we're not using that anywhere AFAIK) and that we need to save the `type` of each metadata value for frontend display purposes. Perhaps we could combine all the values with similar foreign keys into one big dict that maps from keys to values and types but IMO that seems quite convoluted so I actually prefer the current implementation.

Reply:
Now that I think about it, the API part is actually really poorly implemented:

```python
for key, value in metadata.items():
    metadata_model = self.zen_store.create_run_metadata(run_metadata)
```

This even does N *sequential* API calls. Really really bad. Not sure why I would have ever thought this is fine, will look into optimizing this a bit 👍

Reply:
Good catch 👍🏼

Reply:
I rewrote `create_run_metadata()` of the client and all stores now to make batched API requests. That way we at least only make one request per artifact.